src/main/frontend/pipeline-console-view/pipeline-console/main/DataTreeView.tsx

```diff
@@ -76,23 +45,37 @@ const getTreeItemsFromStepList = (stepsItems: StepInfo[]) => {
         key={stepItemData.id}
         nodeId={String(stepItemData.id)}
         label={stepItemData.name}
+        stageId={String(stepItemData.stageId)}
       />
     );
   });
 };
 
 const getTreeItemsFromStage = (
   stageItems: StageInfo[],
-  stageSteps: Map<string, StepInfo[]>
+  steps: StepInfo[],
 ) => {
+  // Copy steps so we don't affect props.steps.
+  let stepsCopy = [...steps];
   return stageItems.map((stageItemData) => {
     let children: JSX.Element[] = [];
+    let stageSteps = [] as StepInfo[];
+    // Handle leaf nodes first.
     if (stageItemData.children && stageItemData.children.length > 0) {
-      children = [...getTreeItemsFromStage(stageItemData.children, stageSteps)];
+      children = getTreeItemsFromStage(stageItemData.children, stepsCopy);
     }
-    let steps = stageSteps.get(`${stageItemData.id}`);
-    if (steps) {
-      let stepsItems = getTreeItemsFromStepList(steps);
+    var i = stepsCopy.length
+    while (i--) {
+      let step = stepsCopy[i];
+      if (step.stageId == String(stageItemData.id)) {
+        stageSteps.push(step);
+        // Remove step from local copy - can only have one parent.
+        // This should reduce the total number of loops required.
+        stepsCopy.splice(i, 1);
+      }
```

Comment:
not sure if this is related but it would be good if nested parents worked right,

e.g. parallel(linux-11) -> stage(checkout(linux-11))

for an example of where it doesn't work see:
https://ci.jenkins.io/job/Plugins/job/slack-plugin/job/master/325/pipeline-console/?selected-node=25

Reply:
I'm not sure I follow this.
This should only be affecting steps, and (I think) steps should only have one parent.

It can be removed - it's an optimisation, not a functional component.

Reply:
(this was broken before, wasn't sure if this code was related, I can raise a separate issue for it anyway)

Reply:
raised https://github.com/jenkinsci/pipeline-graph-view-plugin/issues/92